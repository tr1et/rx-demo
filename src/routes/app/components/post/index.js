export { Post } from './post.component';
