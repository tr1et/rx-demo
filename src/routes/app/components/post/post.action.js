import { dispatcher } from 'shared/services';

import { changeLikes } from '../../app.action';

export const postRxActions = { changeLikes: dispatcher(changeLikes) };
