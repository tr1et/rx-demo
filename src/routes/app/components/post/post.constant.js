// Constants
export const LIKE_STREAK_DEBOUNCE_TIME = 300;
export const LIKE_STREAK_START_THRESHOLD = 10;
export const LIKE_STREAK_MULTIPLY_FACTOR = 10;
export const LIKE_STREAK_SPAN_MINIMUM_ZOOM = 1;
export const LIKE_STREAK_SPAN_MAXIMUM_ZOOM = 25;
