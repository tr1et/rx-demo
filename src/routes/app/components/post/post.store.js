import { likes } from '../../app.store';

export const postRxState = { likes };
