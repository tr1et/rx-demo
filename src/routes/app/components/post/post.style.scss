.post {
  &__streak {
    font-family: 'Lobster Two', cursive;
    transition: transform 500ms ease;
    text-shadow: 0 0 10px white;
  }

  &__like-button,
  &__dislike-button {
    padding: 15px 20px;
    margin: -15px 0;
  }
}
