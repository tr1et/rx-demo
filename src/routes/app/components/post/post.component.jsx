import { Component, Fragment } from 'react';
import { string, number, func, objectOf, instanceOf } from 'prop-types';
import { Card, Icon, Avatar, Progress } from 'antd';
import { Observable, Subject, BehaviorSubject } from 'rxjs';
import { clamp } from 'lodash';

import { rxBind } from 'shared/utilities';

import { postRxActions } from './post.action';
import { postRxState } from './post.store';
import {
  LIKE_STREAK_DEBOUNCE_TIME,
  LIKE_STREAK_DECREASE_TIME,
  LIKE_STREAK_START_THRESHOLD,
  LIKE_STREAK_MULTIPLY_FACTOR,
  LIKE_STREAK_SPAN_MINIMUM_ZOOM,
  LIKE_STREAK_SPAN_MAXIMUM_ZOOM,
} from './post.constant';
import './post.style.scss';

export class Post extends Component {
  static propTypes = {
    poster: string.isRequired,
    avatar: string.isRequired,
    title: string.isRequired,
    message: string,
    maxLikes: number.isRequired,
    rxState: objectOf(instanceOf(Observable)),
    rxActions: objectOf(func),
  };

  static defaultProps = {
    message: '',
    rxState: postRxState,
    rxActions: postRxActions,
  };

  constructor(props, context) {
    super(props, context);

    this.state = {
      likes: 0,
      streak: 0,
      buffer: 0,
    };
    this.onIncrease = rxBind();
    this.onDecrease = rxBind();
    this.unsubscribe$ = new Subject();
  }

  /**
   * Setup observables when component mount
   *
   * @returns {void}
   *
   * @memberOf Post
   */
  componentWillMount() {
    // The changes to `likes` multiplied with streak factor
    const multipliedChangesToLikesState$ = this.likesOrDislikes$.withLatestFrom(
      this.likeStreaks$,
      this.getMultipliedChangeOfLikes,
    );

    // Start a buffer of changes to `likes` state at the beginning
    // and reset everytime the streak end
    const bufferedChangesToLikesState$ = Observable.of('START')
      .merge(this.stopLikesOrDislikes$)
      .switchMap(() => multipliedChangesToLikesState$.scan((buffer, change) => buffer + change, 0))
      .share();

    // Dispatch the buffered `likes` changes to Rx store when the streak end
    // Used to reduce update requests to Rx state.
    bufferedChangesToLikesState$
      .audit(() => this.stopLikesOrDislikes$)
      .subscribe(this.props.rxActions.changeLikes);

    // Update component state based on Rx `likes` state, the like streak and the current buffered
    // change
    Observable
      .merge(
        this.rxStateLikes$.map(likes => ({ likes })),
        this.likeStreaks$.map(streak => ({ streak })),
        bufferedChangesToLikesState$.merge(this.stopLikesOrDislikes$).map(buffer => ({ buffer })),
      )
      .scan((state, newState) => ({ ...state, ...newState }), {})
      .debounceTime(0)
      .subscribe(newState => this.setState(newState));
  }

  /**
   * Unsubscribe all observable when the component unmount
   *
   * @returns {void}
   *
   * @memberOf Post
   */
  componentWillUnmount() {
    this.unsubscribe$.next();
    this.unsubscribe$.complete();
  }

  /**
   * The observable of `likes` state in Rx state
   *
   * @readonly
   *
   * @memberOf Post
   */
  get rxStateLikes$() {
    return this.props.rxState.likes$.takeUntil(this.unsubscribe$);
  }

  /**
   * The observable of changes to `likes` state, includes both "like" and "dislike".
   * It doesn't actually changes the `likes` state
   *
   * @readonly
   *
   * @memberOf Post
   */
  get likesOrDislikes$() {
    // The observable of `like` click, map to the amount of adding likes
    const like$ = this.onIncrease.$.takeUntil(this.unsubscribe$).mapTo(1);

    // The observable of `dislike` click, map to the amount of minusing likes
    const dislike$ = this.onDecrease.$.takeUntil(this.unsubscribe$).mapTo(-1);

    // The changes to `likes` state, combine `like` and `dislike`
    return Observable.merge(like$, dislike$);
  }

  /**
   * The observable of the ends of like streak
   * Emits when the the user stop like/dislike for a specified time
   *
   * @readonly
   *
   * @memberOf Post
   */
  get stopLikesOrDislikes$() {
    if (!this._stopLikesOrDislikes$) {
      this._stopLikesOrDislikes$ = this.likesOrDislikes$
        .debounceTime(LIKE_STREAK_DEBOUNCE_TIME)
        .mapTo(0)
        .share();
    }

    return this._stopLikesOrDislikes$;
  }

  /**
   * Observable of like streaks.
   * A like streak is is a range of likes/dislikes in a short time
   *
   * @readonly
   *
   * @memberOf Post
   */
  get likeStreaks$() {
    if (!this._likeStreaks$) {
      const likesObserver$ = new BehaviorSubject(0);
      const likeStreak$ = likesObserver$.scan((streak, value) => streak + value, 0).share();
      const likeStreakIsEmpty$ = likeStreak$.filter(streak => streak === 0);

      // The likes/dislikes will increase the streak
      this.likesOrDislikes$.mapTo(1).subscribe(likesObserver$);

      // The streak will decrease gradually when user stop like/dislike
      this.stopLikesOrDislikes$
        .switchMap(() => Observable
          .interval(LIKE_STREAK_DECREASE_TIME)
          .mapTo(-1)
          .takeUntil(likeStreakIsEmpty$.merge(this.likesOrDislikes$)))
        .subscribe(likesObserver$);

      this._likeStreaks$ = likeStreak$;
    }

    return this._likeStreaks$;
  }

  /**
   * Get the multiplied change of `likes` state.
   * The multiply factor starts at 1 and double for every 10 likes/dislikes in the current like
   * streak.
   *
   * @param {number} change The original change to `likes` state
   * @param {number} streak The current streak score
   * @returns {number} Multiplied change
   *
   * @memberOf Post
   */
  getMultipliedChangeOfLikes = (change, streak) =>
    change * (2 ** Math.floor(streak / LIKE_STREAK_MULTIPLY_FACTOR));

  /**
   * Get the zoom level of the like streak number, with lower and upper cap
   *
   * @param {number} streak The current streak count
   * @returns {number} The zoom level of the span
   * @memberOf Post
   */
  getLikeStreakZoom = streak =>
    clamp(
      Math.floor(streak / LIKE_STREAK_MULTIPLY_FACTOR),
      LIKE_STREAK_SPAN_MAXIMUM_ZOOM - LIKE_STREAK_SPAN_MINIMUM_ZOOM,
    ) + LIKE_STREAK_SPAN_MINIMUM_ZOOM;

  /**
   * Render the center span with present the likes number or the like streak
   *
   * @returns {Element}
   *
   * @memberOf Post
   */
  renderCenterSpan() {
    const { likes, streak, buffer } = this.state;

    if (streak < LIKE_STREAK_START_THRESHOLD) {
      return <span className="post__likes-count center">+{likes + buffer}</span>;
    }

    const likeStreakZoom = this.getLikeStreakZoom(streak);
    const likeStreakColor = buffer ? 'green' : 'red';

    return (
      <span>
        <span
          className={`post__streak z2 center absolute inline-block ${likeStreakColor}`}
          style={{ transform: `scale(${likeStreakZoom}) translateY(-1em)` }}
        >
          x{streak}
        </span>
        <span className="post__likes-count center">+{likes + buffer}</span>
      </span>
    );
  }

  /**
   * Render the component
   *
   * @returns {Element}
   *
   * @memberOf Post
   */
  render() {
    const { likes, buffer } = this.state;
    const { poster, avatar, title, message, maxLikes } = this.props;
    const percent = clamp(((likes + buffer) / maxLikes) * 100, 100);

    return (
      <Fragment>
        <Card
          cover={<img alt="poster" src={poster} />}
          actions={[
            <Icon className="post__like-button" type="like-o" onClick={this.onIncrease} />,
            this.renderCenterSpan(),
            <Icon className="post__dislike-button" type="dislike-o" onClick={this.onDecrease} />,
          ]}
        >
          <Card.Meta
            avatar={<Avatar size="large" src={avatar} />}
            title={title}
            description={message}
          />
        </Card>
        <Progress percent={percent} showInfo={false} />
      </Fragment>
    );
  }
}
