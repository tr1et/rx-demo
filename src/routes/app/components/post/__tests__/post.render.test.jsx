import React from 'react';
import { mount } from 'enzyme';
import { rxSandbox } from 'rx-sandbox';
import faker from 'faker';
import delay from 'promise-delay';

import { Post } from '../post.component';

faker.seed(1234);

const defaultProps = {
  poster: faker.image.imageUrl(),
  avatar: faker.image.imageUrl(),
  title: faker.name.findName(),
  message: faker.lorem.sentence(),
  maxLikes: 1000,
};
let marbles = null;

const normalStateTest = () => {
  test('HTML of component in normal state', async () => {
    // arrange
    const rxState = { likes$: marbles.cold('a', { a: 120 }) };
    const props = { ...defaultProps, rxState };

    expect.assertions(2);

    // act
    const wrapper = mount(<Post {...props} />);
    marbles.flush();
    await delay(0, 0);

    const likesCountElement = wrapper.find('.post__likes-count');
    const html = wrapper.html();

    // assert
    expect(likesCountElement.text()).toBe('+120');
    expect(html).toMatchSnapshot();
  });
};

const likeStreakStateTest = () => {
  test('HTML of component in like streak state', async () => {
    // arrange
    const rxState = { likes$: marbles.cold('a', { a: 120 }) };
    const props = { ...defaultProps, rxState };
    const state = {
      streak: 25,
      buffer: 80,
    };

    expect.assertions(4);

    // act
    const wrapper = mount(<Post {...props} />);
    marbles.flush();
    wrapper.setState(state);
    await delay(0, 0);

    const likesCountElement = wrapper.find('.post__likes-count');
    const likeStreakElement = wrapper.find('.post__streak');
    const html = wrapper.html();

    // assert
    expect(likesCountElement.text()).toBe('+200');
    expect(likeStreakElement.text()).toBe('x25');
    expect(likeStreakElement.prop('style')).toEqual({ transform: 'scale(3) translateY(-1em)' });
    expect(html).toMatchSnapshot();
  });
};

describe('post.snapshot.test.js', () => {
  beforeEach(() => {
    marbles = rxSandbox.create();
  });

  normalStateTest();
  likeStreakStateTest();
});
