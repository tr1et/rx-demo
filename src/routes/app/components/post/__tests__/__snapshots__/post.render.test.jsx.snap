// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post.snapshot.test.js HTML of component in like streak state 1`] = `

<div class="ant-card ant-card-bordered">
  <div class="ant-card-cover">
    <img alt="poster"
         src="http://lorempixel.com/640/480"
    >
  </div>
  <div class="ant-card-body">
    <div class="ant-card-meta">
      <div class="ant-card-meta-avatar">
        <span class="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-image">
          <img src="http://lorempixel.com/640/480">
        </span>
      </div>
      <div class="ant-card-meta-detail">
        <div class="ant-card-meta-title">
          Lilly Schumm III
        </div>
        <div class="ant-card-meta-description">
          Repellendus est dolor eos enim voluptatem ad.
        </div>
      </div>
    </div>
  </div>
  <ul class="ant-card-actions">
    <li style="width: 33.333333333333336%;">
      <span>
        <i class="anticon anticon-like-o post__like-button">
        </i>
      </span>
    </li>
    <li style="width: 33.333333333333336%;">
      <span>
        <span>
          <span class="post__streak z2 center absolute inline-block green">
            x25
          </span>
          <span class="post__likes-count center">
            +200
          </span>
        </span>
      </span>
    </li>
    <li style="width: 33.333333333333336%;">
      <span>
        <i class="anticon anticon-dislike-o post__dislike-button">
        </i>
      </span>
    </li>
  </ul>
</div>

`;

exports[`post.snapshot.test.js HTML of component in normal state 1`] = `

<div class="ant-card ant-card-bordered">
  <div class="ant-card-cover">
    <img alt="poster"
         src="http://lorempixel.com/640/480"
    >
  </div>
  <div class="ant-card-body">
    <div class="ant-card-meta">
      <div class="ant-card-meta-avatar">
        <span class="ant-avatar ant-avatar-lg ant-avatar-circle ant-avatar-image">
          <img src="http://lorempixel.com/640/480">
        </span>
      </div>
      <div class="ant-card-meta-detail">
        <div class="ant-card-meta-title">
          Lilly Schumm III
        </div>
        <div class="ant-card-meta-description">
          Repellendus est dolor eos enim voluptatem ad.
        </div>
      </div>
    </div>
  </div>
  <ul class="ant-card-actions">
    <li style="width: 33.333333333333336%;">
      <span>
        <i class="anticon anticon-like-o post__like-button">
        </i>
      </span>
    </li>
    <li style="width: 33.333333333333336%;">
      <span>
        <span class="post__likes-count center">
          +120
        </span>
      </span>
    </li>
    <li style="width: 33.333333333333336%;">
      <span>
        <i class="anticon anticon-dislike-o post__dislike-button">
        </i>
      </span>
    </li>
  </ul>
</div>

`;
